docs/x509.rst

```diff
@@ -468,6 +468,73 @@ X.509 Revoked Certificate Object
 
         The extensions encoded in the revoked certificate.
 
+X.509 CSR (Certificate Signing Request) Builder Object
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. class:: CertificateSigningRequestBuilder
+
+    .. versionadded:: 1.0
+
+    .. doctest::
+
+        >>> from cryptography import x509
+        >>> from cryptography.hazmat.backends import default_backend
+        >>> from cryptography.hazmat.primitives import hashes
+        >>> from cryptography.hazmat.primitives.asymmetric import rsa
+        >>> private_key = rsa.generate_private_key(
+        ...     public_exponent=65537,
+        ...     key_size=2048,
+        ...     backend=default_backend()
+        ... )
+        >>> builder = x509.CertificateSigningRequestBuilder()
+        >>> builder = builder.set_subject_name(x509.Name([
+        ...     x509.NameAttribute(x509.OID_COMMON_NAME, 'cryptography.io'),
+        ... ]))
+        >>> buidlder = builder.add_extension(x509.Extension(
+        ...     x509.OID_BASIC_CONSTRAINTS,
+        ...     True,
+        ...     x509.BasicConstraints(False, None),
+        ... ))
+        >>> request = builder.sign(
+        ...     default_backend(), private_key, hashes.SHA1()
+        ... )
+        >>> isinstance(request, x509.CertificateSigningRequest)
+        True
+
+    .. method:: __init__()
+
+        Creates an empty certificate signing request.
+
+    .. method:: set_subject_name(name)
+
+        :param name: The :class:`Name` of the certificate subject.
+        :returns: A new `CertificateSigningRequestBuilder`.
+
+    .. method:: add_extension(extension)
```

Comment:
That sounds reasonable. It would preclude this API being used to do custom extensions, but that's not possible right now anyway since the underlying backend needs to know how to serialize the extension (and we have no plans to allow users to define arbitrary serializers).